### PR TITLE
Onedrive exporter

### PR DESCRIPTION
**Pull request by liviasalgueiro**
Description:
Create a storage mechanism to save feeds in OneDrive (any format) as discussed in https://zyte.eu.teamwork.com/app/tasks/24691756

### CODE REVIEW

File: scrapy_onedrive_exporter/onedrive_exporter.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            raise NotConfigured(
+                "OneDriveFeedStorage does not support appending to files. Please "
+                "remove the overwrite option from your FEEDS setting or set it to True."
+            )
+
+        self.session_headers = {
+            'Authorization': f'Bearer {access_token}',
+            'Content-Type': 'application/json'
+        }
+        self.file_path = uri.replace('onedrive://', '')
```
Comment (Gallaecio) [inline scrapy_onedrive_exporter/onedrive_exporter.py]: Let’s not assume which scheme the user chose (e.g. maybe they went for `od`).

File: scrapy_onedrive_exporter/onedrive_exporter.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            access_token=crawler.settings.get("ONEDRIVE_ACCESS_TOKEN"),
+            feed_options=feed_options,
+        )
+
+    def _create_session(self, file_path):
+        session_endpoint = f"/me/drive/root:/{file_path}:/createUploadSession"
+        response = requests.post(GRAPH_API_ENDPOINT + session_endpoint, headers=self.session_headers)
+        try:
+            return response.json()['uploadUrl']
+        except Exception as e:
```
Comment (Gallaecio) [inline scrapy_onedrive_exporter/onedrive_exporter.py]: Can we limit this to `KeyError`, or are there other possible exceptions?

File: scrapy_onedrive_exporter/onedrive_exporter.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            raise NotConfigured(
+                f"Failed to create the upload Session: {e}. Response: {response.content}"
+            )
+
+    def _store_in_thread(self, file):
+        file.seek(0)
+        content = file.read()
+        content_range = f"bytes 0-{len(content) - 1}/{len(content)}"
+        headers = {'Content-Length': str(len(content)), 'Content-Range': content_range}
```
Comment (Gallaecio) [inline scrapy_onedrive_exporter/onedrive_exporter.py]: Do we need to specify `Content-Length`? I would expect `requests` to set that one automatically, although maybe it does not if you set `Content-Range` as well.

Comment (liviasalgueiro) [reply scrapy_onedrive_exporter/onedrive_exporter.py]: @Gallaecio I tried removing it and it worked, please check again. thanks!

File: scrapy_onedrive_exporter/onedrive_exporter.py (hunk 2 above)
Comment (liviasalgueiro) [reply scrapy_onedrive_exporter/onedrive_exporter.py]: the only other I can think of is a JSONdecoderError if the response is not JSON.. I added that, lmk please

File: scrapy_onedrive_exporter/onedrive_exporter.py (hunk 1 above)
Comment (liviasalgueiro) [reply scrapy_onedrive_exporter/onedrive_exporter.py]: @Gallaecio the documentation mentions the scheme should be onedrive, but I updated to handle others as well. Let me know

File: scrapy_onedrive_exporter/onedrive_exporter.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            raise NotConfigured(
+                "OneDriveFeedStorage does not support appending to files. Please "
+                "remove the overwrite option from your FEEDS setting or set it to True."
+            )
+
+        self.session_headers = {
+            'Authorization': f'Bearer {access_token}',
+            'Content-Type': 'application/json'
+        }
+        self.file_path = uri.split("://")[-1]
```
Comment (Gallaecio) [inline scrapy_onedrive_exporter/onedrive_exporter.py]:
```suggestion
        self.file_path = uri.split("://", max_split=1)[-1]
```

(`https://example.com/://` is a valid URL)

**Review by Gallaecio — APPROVED** [review]
Great job!

File: README.md [hunk 5]
```diff
@@ -1 +1,51 @@
...
...6BAAUs5%2bHQn0N%2bh2FxWzLS31ZgQVuHsYAAQe0m4RTofw7c6jCDsPCN5Hq6RYwdPsQxxJZKLJBFXt8SCoSJ3BeN36l/YoTBHaBNjyI8KmYtr1AcYx93wqvojL4g1PNR%2bLRqFSzQe2PyhumzMsGGbFqmZzNtzluaQSm3rdtSorbriPL3AFuWcIpo0AnD7iS666OLolARAnCpqLHY2sNyM6SUIo3DZgLEUrJwS6S05YWAQQvjuNiQsSG49r8jmHJ6O9cDbsXXjpTku3UtSN3558cuOordIj5mQQ4evJ1dDPRI3L18%2bTgilmAqOcz6R%2b%2b5YhUk3cU854L9gZe86WsRjfs9ztTvwc3IpQD4ICoJ94L1kNGZjlKRQBfBsMDZgAACF9C2ej1pyckOALbj9o2vGrEU/7DpKSmjCyf7IaVcbzohSH1exYk4qmovogcIcHx%2bqjklQyosJsHt5k1RzXg69GS/9hde9h%2blBSx3d30ceDO309jrTm2S9lG9Qe5PeYA6G062Ros0ms%2btfxmL3NJTz/eIWWIYICoFjpd85WGTpG1qAWvF6UY%2bJZbME2x8ewM9oVGQfHY4WDWcquMEU/UYpdlxlV3RHcv/tqKoqR1jqJE6A45fo7u%2bPGfBlg7HQktVZ%2bnT6gK054xq6dbEGByRunZdGHZs3ThpC40wCcOHzuK7Z9GVhbnRDIeMGCPofCVkn/z7uuNgowAEAadfOZyuGRLTwNrTiG70mo5kub5o3tX5lLUdrkbgbK6riBK8rsN6tlZUENymMU95A8svBqlsm27fxDk2D/S2ltEmQLBlULKXZPmpcIfHUNAoum%2bhfVPQZoN5nzVa3d/7RYDT0S4%2bd/5jsV5yoem0/Yd8VnclvJs14UC4Eq%2bw7L3r314rAMojg5rgbMhfmei2PharkNctvn%2b5j8Ay35YbFVknhLrj1lY5I1GcCX4J5qADbdn8l%2b1fg3B3H3wVV9yo1EjFaRW4LObrOwTcjQRPkC4FMbDbpSitXeRaSiNrNZuVzSXgUcHIcGTUXsYz89c7n55bHQDKLtWThXT%2bUEgNR/8%2bIqpdBukM1S78h4EOZo8mz5sj%2b7znPRAqyCedrLLL5bleq2viUCx3dFThnQ6ChxV64E508C96VACj4PofHpG2sfB6PJIrP0MfgI%3d'
```
Comment (VMRuiz) [inline README.md:34]: Please confirm this is a fake token 😄

Comment (Gallaecio) [reply README.md:34]: Even if it isn’t, it might have expired by now :crossed_fingers:

Comment (liviasalgueiro) [reply README.md:34]: oh yes, it's a random invalid token :)

**Review by VMRuiz — APPROVED** [review]

**Comment by VMRuiz** [comment]
I think we are good to merge. Who has permissions for that?

**Pull request merged**